### PR TITLE
Final reader class implementation

### PR DESCRIPTION
**Pull request by Karsten1987**
According to the comment [here](https://github.com/ros2/rosbag2/pull/191#discussion_r345423931), that's how I would envision the composition to happen. 

The basic reader class is `final` and it's essentially dependency injection which decides on which reader implementation to use for reading the bag file.

### CODE REVIEW

**Comment by Karsten1987** [comment]
I believe the same pattern should be applied to the writer class

**Comment by prajakta-gokhale** [comment]
How do you envision unit testing to change when the `Reader` class needs to be mocked? Like in rosbag2_transport [mock_sequential_reader](https://github.com/ros2/rosbag2/blob/master/rosbag2_transport/test/rosbag2_transport/mock_sequential_reader.hpp) for example.

**Comment by Karsten1987** [comment]
Does the second commit on this PR makes sense? https://github.com/aws-ros-dev/rosbag2/pull/4/commits/e9be36b6e24fe3bb0e82799a5b4a7d57c96b63ec

Or maybe I understand your question incorrectly. The way I see this is that the actual Reader instance should never be mocked as it basically doesn't perform any work. And so the beauty of this is that mocking the actual sequential reader implementation is now trivial as it's just another implementation of the base reader interface.

Does this make sense?

**Comment by prajakta-gokhale** [comment]
> Does the second commit on this PR makes sense? [e9be36b](https://github.com/aws-ros-dev/rosbag2/commit/e9be36b6e24fe3bb0e82799a5b4a7d57c96b63ec)
> 
> Or maybe I understand your question incorrectly. The way I see this is that the actual Reader instance should never be mocked as it basically doesn't perform any work. And so the beauty of this is that mocking the actual sequential reader implementation is now trivial as it's just another implementation of the base reader interface.
> 
> Does this make sense?

Yes, it does. I didn't see the second commit when I wrote that comment earlier. I think this is the right approach now :)

File: rosbag2_transport/test/rosbag2_transport/test_play.cpp [hunk 1]
```diff
@@ -19,6 +19,7 @@
 #include <memory>
 #include <string>
 #include <vector>
+#include <utility>
```
Comment (prajakta-gokhale) [inline rosbag2_transport/test/rosbag2_transport/test_play.cpp:22]: What is this needed for?

Comment (Karsten1987) [reply rosbag2_transport/test/rosbag2_transport/test_play.cpp:22]: for the call to `std::make_unique`

Comment (prajakta-gokhale) [reply rosbag2_transport/test/rosbag2_transport/test_play.cpp:22]: I believe that should be covered in `<memory>` per https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique.

Comment (Karsten1987) [reply rosbag2_transport/test/rosbag2_transport/test_play.cpp:22]: my bad, it's needed for `std::move`. 
https://en.cppreference.com/w/cpp/utility/move

I mean, cppcheck was complaining about it in the first place ;-)

**Review by prajakta-gokhale — APPROVED** [review]
LGTM. @thomas-moulard could you take a look?

File: rosbag2/include/rosbag2/reader.hpp [hunk 2]
```diff
@@ -43,22 +43,17 @@ namespace rosbag2
...
-class ROSBAG2_PUBLIC SequentialReader
+class ROSBAG2_PUBLIC Reader final
 {
 public:
-  explicit SequentialReader(
-    std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory =
-    std::make_unique<rosbag2_storage::StorageFactory>(),
-    std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory =
-    std::make_shared<SerializationFormatConverterFactory>());
+  Reader(std::unique_ptr<BaseReaderInterface> reader_impl);
```
Comment (thomas-moulard) [inline rosbag2/include/rosbag2/reader.hpp]:
```suggestion
  explicit Reader(std::unique_ptr<BaseReaderInterface> reader_impl);
```

File: rosbag2/include/rosbag2/reader.hpp [hunk 3]
```diff
@@ -43,22 +43,17 @@ namespace rosbag2
 class BaseReaderInterface;
 
 /**
- * The SequentialReader allows opening and reading messages of a bag. Messages will be read
- * sequentially according to timestamp.
+ * The Reader allows opening and reading messages of a bag.
```
Comment (thomas-moulard) [inline rosbag2/include/rosbag2/reader.hpp:49]: any reason to change this comment? The old version seems more precise?

File: rosbag2/include/rosbag2/reader.hpp [hunk 4]
```diff
@@ -43,22 +43,17 @@ namespace rosbag2
...
-    std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory =
-    std::make_shared<SerializationFormatConverterFactory>());
+  Reader(std::unique_ptr<BaseReaderInterface> reader_impl);
 
-  virtual ~SequentialReader();
+  ~Reader();
 
   /**
-   * Open a rosbag for reading messages sequentially (time-ordered). Throws if file could not be
+   * Throws if file could not be
```
Comment (thomas-moulard) [inline rosbag2/include/rosbag2/reader.hpp]: could you break the line at the end of the first sentence?

File: rosbag2/include/rosbag2/readers/sequential_reader.hpp [hunk 5]
```diff
@@ -29,18 +29,21 @@
...
 #include "rosbag2/visibility_control.hpp"
 
-
 namespace rosbag2
 {
+namespace readers
+{
 
-class ROSBAG2_LOCAL SequentialReaderImpl : public BaseReaderInterface
+class ROSBAG2_PUBLIC SequentialReader : public BaseReaderInterface
```
Comment (thomas-moulard) [inline rosbag2/include/rosbag2/readers/sequential_reader.hpp]: Is sequential reader meant to be instantiated outside the library? It would be preferable to keep headers private so that we can keep updating rosbag2 w/o breaking ABI compat.

File: rosbag2/include/rosbag2/readers/sequential_reader.hpp [hunk 6]
```diff
@@ -29,18 +29,21 @@
...
 {
 public:
-  SequentialReaderImpl(
-    std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory,
-    std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory);
+  SequentialReader(
+    std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory =
+    std::make_unique<rosbag2_storage::StorageFactory>(),
+    std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory =
+    std::make_shared<SerializationFormatConverterFactory>());
```
Comment (thomas-moulard) [inline rosbag2/include/rosbag2/readers/sequential_reader.hpp:44]: Is this change necessary?

**Review by thomas-moulard — COMMENTED** [review]
I missed a lot of design discussion so I don't have the appropriate context, but moving private headers to make them public is a dangerous slope.

Can't we implement the same thing w/o making `SequentialReader` public?

**Comment by Karsten1987** [comment]
To summarize:

The `Reader` class is essentially nothing else than a wrapper around individual reader implementations inheriting from `BaseReaderClass`. That gives the flexibility to instantiate a reader with different behavior. So for example, that allows to easily replace a time sequential reader implementation to be replaced by something completely else without breaking API, even ABI as simply a new pointer to interface will be passed into the instantiation of the `Reader`. Hence, the `Reader` class can even be final. 

That's the way at least I understood your requirement of being flexible to adapt to new features. So that would allow to instantiate e.g. a reader capable of reading through `bz2` compressed data, simply by adding a new `CompressedSequentialReader` class inheriting from the same base interface class. Which in fact also means no break of API/ABI as the new reader implementation can even live outside of the rosbag2 repo itself. The implementation of the reader could e.g. live in its own `rosbag2_compression` package, which provides the algorithm to compress messages and such a reader implementation to extract the messages accordingly.

As I mentioned in a [previous comment](https://github.com/ros2/rosbag2/pull/191#pullrequestreview-305611310), I am also totally cool with having the reader class strictly pimpl'ed, but I don't see how that would allow to easily replace the logic of the reader implementation to cope with compression or non time sequenced replay behavior in the future.
Essentially, calls like this: https://github.com/ros2/rosbag2/pull/191/files#diff-5c7c678bca1d3e21c17d1e55dcdd6eefR31 as here the `SequentialReaderImpl` is hardcoded.

> I missed a lot of design discussion so I don't have the appropriate context, but moving private headers to make them public is a dangerous slope.

The original PR moved the Sequential Reader to a public include: https://github.com/ros2/rosbag2/pull/191/files#diff-306e9ce6e9ceac4de3023adf9030022e
Which, to be fair, makes a lot of sense to me so that variations of this can easily inherit from an existing reader implementation and don't have to implement the base interface from scratch.

**Comment by thomas-moulard** [comment]
I see, that makes sense. Let's go with this solution then.

File: rosbag2/include/rosbag2/reader.hpp (hunk 2 above)
Comment (Karsten1987) [reply rosbag2/include/rosbag2/reader.hpp]: addressed in https://github.com/aws-ros-dev/rosbag2/pull/4/commits/5a670023464e57f2c12a2379b0bac2758cdf7e1d

File: rosbag2/include/rosbag2/reader.hpp (hunk 3 above)
Comment (Karsten1987) [reply rosbag2/include/rosbag2/reader.hpp:49]: as per discussion

File: rosbag2/include/rosbag2/reader.hpp (hunk 4 above)
Comment (Karsten1987) [reply rosbag2/include/rosbag2/reader.hpp]: addressed in https://github.com/aws-ros-dev/rosbag2/pull/4/commits/5a670023464e57f2c12a2379b0bac2758cdf7e1d

**Comment by Karsten1987** [comment]
addressed comments and rebased on top of `reader-interface` branch.
